File: programmers/contest/피로도.cpp
// https://programmers.co.kr/learn/courses/30/lessons/87946

/*
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int need[8], use[8];
bool check[8];
int num, res = 0;

void dfs(int cnt, int remain) {
    for (int i = 0; i < num; i++) {
        if (!check[i] && need[i] <= remain) {
            check[i] = true;
            dfs(cnt + 1, remain - use[i]);
            check[i] = false;
        }
    }
    res = max(res, cnt);
    return;
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    num = dungeons.size();
    for (int i = 0; i < num; i++) {
        need[i] = dungeons[i][0];
        use[i] = dungeons[i][1];
    }
    dfs(0, k);
    return answer = res;
}

*/

File: programmers/피로도.py
# 탐험을 시작하기 위한 최소필요 피로도, 던전탐험을 마쳤을때 소모되는 소모피로도가 있음
# 게임에 하루에 한번씩 탐험할수있는 던전이 여러개있음
# 유저가 탐험할수있는 최대 던전 수??
def solution(k, dungeons):
    answer = 0
    n = len(dungeons)
    visited = [False] * n

    def dfs(hp, count):
        nonlocal answer
        answer = max(answer, count)
        for index, [need_hp, use_hp] in enumerate(dungeons):
            if visited[index] or hp < need_hp:
                continue
            visited[index] = True
            dfs(hp - use_hp, count + 1)
            visited[index] = False

    dfs(k, 0)
    return answer

